Enterprise-CMCS/macpro-mdct-qmr
# Refactor Combined Rates

### Description
There are a ton of changes in here, but the driving force is these three big ones:

**1. Fix spec misunderstanding.** The original implementation was partly based on a miscommunication of the requirements - that only specific combinations of data source selections could trigger a combined rate calculation. This required tracking a lot of details that turned out to be unnecessary. It's a beautiful thing when the more technically correct way happens to be the simpler way.

**2. Compute rates more directly.** The original implementation performed a series of successive transformations on the input measure data, some of which took place on the frontend. This implementation builds (nearly) display-ready objects straight from the inputs. In some cases this reduces code; fewer transformations are necessary. In other cases it increases code; some arrays with maps have become individual variables in the code.

**3. Separate Rates from Other Things.** Certain measures have, in addition to the typical "numerator, denominator, rate" objects, some non-rate values that still need combining (for example, CPU-AD's count of individuals unreachable for care planning). These values are stored mixed-in with the typical rates, and the original implementation kept them mixed in. Dealing with these values (and their differences from typical rates) will always be complex, but this implementation draws a clear line to hide that complexity away (for the vast majority of measures).

I have left additional comments on the changed files.

### Related ticket(s)
* CMDCT-3651 - Combine rates for admin+admin. Functionality rewritten in this branch.
* CMDCT-3652 - Combine rates for hybrid+hybrid. Functionality rewritten in this branch.
* CMDCT-3653 - Combine rates for hybrid+admin. Functionality rewritten in this branch.
* CMDCT-3877 - Combine values for CPU-AD. Functionality rewritten in this branch.
* CMDCT-3878 - Combine values for PCR-AD. Functionality rewritten in this branch.
* CMDCT-3904 - Use user-overridden rate values when computing weighted combined rate. Bugfix replicated in this branch.
* CMDCT-3903 - Use denominator to weight Hybrid+Admin calculations. Bugfix replicated in this branch.
* CMDCT-3933 - Weighted rate "doesn't recalculate". Bugfix is in this branch only.
* CMDCT-3923 - Display "Not Applicable" for rates that were reported, but cannot be combined. Bugfix is in this branch only.
* CMDCT-3924 - Copy the usable rate to "combined" when one uses an Other data source
* CMDCT-3925 - Copy the usable rate to "combined" when one uses the ECDS data source

---
### How to test
Oh, boy.
1. Log into QMR as a state user, for a state with separated Medicaid & CHIP core sets
2. Fill out some core sets
3. Look at the combined rates.
4. Everything should... work.
5. Least helpful testing instructions ever, I'm sorry.

### Notes
n/a

---
### Pre-review checklist


---
### Pre-merge checklist


#### Review

#### Security
_If either of the following are true, notify the team's ISSO (Information System Security Officer)._
